src/WebJobs.Script.Grpc/Extensions/RpcSharedMemoryDataExtensions.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Script.Grpc.Messages;
+using Microsoft.Azure.WebJobs.Script.Workers.SharedMemoryDataTransfer;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.WebJobs.Script.Grpc.Extensions
+{
+    internal static class RpcSharedMemoryDataExtensions
+    {
+        internal static async Task<RpcSharedMemory> ToRpcSharedMemoryAsync(this object value, ILogger logger, string invocationId, ISharedMemoryManager sharedMemoryManager)
+        {
+            if (sharedMemoryManager.IsSupported(value))
+            {
+                // Put the content into shared memory and get the name of the shared memory map written to
+                SharedMemoryMetadata putResponse = await sharedMemoryManager.PutObjectAsync(value);
+                if (putResponse != null)
+                {
+                    // If written to shared memory successfully, add this shared memory map to the list of maps for this invocation
+                    sharedMemoryManager.AddSharedMemoryMapForInvocation(invocationId, putResponse.Name);
+
+                    RpcSharedMemoryDataType? dataType = GetRpcSharedMemoryDataType(value);
+                    if (dataType.HasValue)
+                    {
+                        // Generate a response
+                        RpcSharedMemory sharedMem = new RpcSharedMemory()
+                        {
+                            Name = putResponse.Name,
+                            Offset = 0,
+                            Count = putResponse.Count,
+                            Type = dataType.Value
+                        };
+
+                        return sharedMem;
+                    }
+                }
+                else
+                {
+                    logger.LogWarning($"Cannot write to shared memory for invocation: {invocationId}");
+                }
+            }
+
+            return null;
+        }
+
+        private static RpcSharedMemoryDataType? GetRpcSharedMemoryDataType(object value)
+        {
+            if (value is byte[])
+            {
```

Comment:
Having this logic here feels a bit disconnected.

If the `IShareMemoryManager` implementation is responsible for defining what is supported, it should handle this as well, right? The scenario where an implementation could add support for additional types, go through the process of having that written,, just to eventually have a `null` response returned because it didn't match the data type mapping here seems wrong. Should we have the type returned by the call to `IsSupported` (perhaps have that renamed to `GetSharedMemoryDataType` with the addition of an `Unknown` or `Unsupported` type to the enumeration) so the manager is providing that as it checks for the type?

Reply:
I somewhat agree with your concern, but then I also don't want to put any RPC concepts (like RPC types i.e. `RpcSharedMemoryDataType` which is specific to Grpc) into the `ISharedMemoryManager` as we don't want to tie it to Grpc or any particular protocol. If we put a `GetSharedMemoryDataType` method in `ISharedMemoryManager` and make it return an `RpcSharedMemoryDataType`, it would tie it to a particular protocol. 
That's why I had decided to d this type conversion between a C# Type and the `RpcSharedMemoryDataType` as part of the `RpcSharedMemoryDataExtensions` as this is specific to Grpc.

Reply:
Indeed, I agree with that. I did miss the fact that this was an RPC specific enumeration (honestly, the Rpc prefixes around the codebase are sometimes a bit confusing in some cases). I still feel like this does introduce a bit of a disconnect. There are a couple of ways we could tackle this, but your approach makes more sense to me with that additional context.